File: graph/keys_and_rooms.py
from typing import List
from collections import deque

class Solution:
    # dfs
    def canVisitAllRooms2(self, rooms: List[List[int]]) -> bool:
        visitSet = set()

        def dfs(key):
            if key in visitSet:
                return
            visitSet.add(key)
            for neiKey in rooms[key]:
                dfs(neiKey)

        dfs(0)
        return len(visitSet) == len(rooms)

    # bfs
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        q = deque([0])
        visitSet = set([0])

        while q:
            key = q.popleft()
            for neiKey in rooms[key]:
                if neiKey not in visitSet:
                    q.append(neiKey)
                    visitSet.add(neiKey)

        return len(visitSet) == len(rooms)

File: graph/keys_and_rooms.java
class Solution {
    public boolean canVisitAllRooms1(List<List<Integer>> rooms) {
        Set<Integer> visited = new HashSet<>();
        dfs(0, visited, rooms);
        if(visited.size() == rooms.size()){
            return true;
        }
        else{
            return false;
        }
    }

    private void dfs(int key, Set<Integer> visited, List<List<Integer>> rooms){
        if(visited.contains(key)){
            return;
        }
        visited.add(key);
        List<Integer> room = rooms.get(key);
        for(int i = 0; i < room.size(); i++){
            dfs(room.get(i), visited, rooms);
        }
    }

    public boolean canVisitAllRooms(List<List<Integer>> rooms){
        Set<Integer> visited = new HashSet<>();
        Deque<Integer> queue = new ArrayDeque<>();
        queue.add(0);

        while(!queue.isEmpty()){
            int key = queue.poll();
            if(visited.contains(key)){
                continue;
            }
            visited.add(key);
            List<Integer> room = rooms.get(key);
            for(int i = 0; i < room.size(); i++){
                queue.add(room.get(i));
            }
        }

        if(visited.size() == rooms.size()){
            return true;
        }
        else{
            return false;
        }
    }
}